Img2img batch processing

breengles opened the pull request:
## Description

* fix the issue for batch img2img processing when the mask file extension might be different from the image file extension so the masks cannot be found.
* add scale_by factor to batch img2img processing which might be helpful if the images are of different sizes/ratios and one would like to preserve them.

## Checklist:

- [x] I have read [contributing wiki page](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing)
- [x] I have performed a self-review of my own code
- [x] My code follows the [style guidelines](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing#code-style)
- [ ] My code passes [tests](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Tests)
well, I tested it and some errors from watermarking occurred. Then I tested the most recent version and the same errors occurred so no new error has been produced :)

breengles committed:
improve filename matching for mask

we should not rely that mask filename will be of the same extension
as the image filename so better pattern matching is added

breengles committed:
add scale_by to batch processing

breengles committed:
ruffed

werran2 commented:
could you fix img2img Batch fuction do not load VAE model?

breengles commented:
> The `**` in the glob seems out of place. `**` is for searching subdirectories.

Indeed, but I use it here purposely as it might provide a more flexible way of folder organization (and should not affect regular usage) if it is not ok for you I will rework it

AUTOMATIC1111 commented:
Do you have an example of how to make use of `**` for this?

breengles commented:
> Do you have an example of how to make use of `**` for this?

suppose you have something like that
```
images
├── <class 1>
    ├── img1.jpg
    ├── img2.jpg
├── <class 2>
    ├── img3.jpg
    ├── img4.jpg

masks
├── <class 1>
    ├── img1.jpg
    ├── img2.jpg
├── <class 2>
    ├── img3.jpg
    ├── img4.jpg
```

and you provide only `path/to/masks`. As I do not want to affect how the `image` is constructed (maybe it also might be useful to support subfolders) I suggest this solution which should work regardless of how you select an image from the `images` folder.

AUTOMATIC1111 commented:
Well, this seems like it would search all masks in the masks dir for every batched image. You should be able to extract `<class 1>/img.jpg` from image filename and append that to `path/to/masks` for full path to image. Also what if you have `img1.jpg` and `img2.jpg` fpor both class 1 and class 2?

breengles commented:
> Well, this seems like it would search all masks in the masks dir for every batched image. You should be able to extract `<class 1>/img.jpg` from image filename and append that to `path/to/masks` for full path to image. Also what if you have `img1.jpg` and `img2.jpg` fpor both class 1 and class 2?

I agree, perhaps better (more precise?) path construction should take place.
Also, I agree that with the current behavior (as it anyway won't find any images in subdirs) of [`shared.listfiles`](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/20ae71faa8ef035c31aa3a410b707d792c8203a3/modules/shared.py#L804) there is no reason in using `**` pattern but we still should not rely on matching extension for images and masks.

breengles commented:
> could you fix img2img Batch fuction do not load VAE model?

@werran2 
Regarding this, it is not clear for me right now why it is happening. In regular img2img everything works fine, right?

breengles committed:
add subdir support for images, masks and output; search mask only in subdir

breengles commented:
@AUTOMATIC1111 Hi!
I reworked the search pattern for masks -- now if there is more than 1 mask available, we search only in the corresponding subdir and take the first one from the list. Nevertheless, if a user has many masks with the same name and different extensions it might be a problem I guess but I do not think it is the real case anyway (either in general there should be only one mask or they should actually be the same). In addition, if no mask has been found in this case the exception is thrown.

Also, I added subdir support for images. masks, and outputs here.

AUTOMATIC1111 commented:
As it is it looks like the code will stop processing once it fails to find mask for one image. Users generally don't want that to happen, they want one image with issues to be skipped and the rest to still be processed.

breengles commented:
> As it is it looks like the code will stop processing once it fails to find mask for one image. Users generally don't want that to happen, they want one image with issues to be skipped and the rest to still be processed.

do we have someway to throw a warning in that case?

breengles commented:
Actually, I found that the proposed retrieval mechanics with subdirs breaks controlnet due to different sorting order of input dirs for `shared.listfiles` (which is used by [sd-webui-conrolnet](https://github.com/Mikubill/sd-webui-controlnet)) and `shared.walk_files` (which I used here). Changing from `shared.listfiles` to `shared.walk_files` in controlnet seems to fix the problem but this should be done simultaneously for both webui itself and the extension. 
Moreover, who knows which extensions also rely on that order so I suggest to fallback one-level dirs here. Will update sooner

breengles committed:
fallback to original file retrieving; skip img if mask not found

usage of `shared.walk_files` breaks controlnet extension
images are processed in different order 
which leads to unmatched img file used for img2img and img file used for controlnet 
(if no folder is specified for control net
or the same as img2img input dir used for it)

DHG-Dav commented:
I just tested the img2img batch, setting the scale factor to 1 (i just want to preserve images sizes/proportions) and it doesn't work for me.
It's scaling the images based on the sliders instead of each image's resolution.
Plus it seems to have a problem with memory garbage collection, since i can't run the code twice if i interrupt it, i run out of memory the second time, unless i relaunch webui. (12Gb vram, working on images under 128X128px in this batch attempt). (only happens in batch img2img using the resizing)
In the commit i saw only /modules/img2img.py changed, so i updated only this file (i had version from a week before this merge anyway), i doubt it's the problem, but i rather mention it, just in case.
Sorry i can't dive into the code and try to figure more, since i'm not working with high level languages.

breengles commented:
Hi @DHG-Dav 
I will take a look… Anyway I should say that in my testings everything works just fine in terms of memory (though, I have 24Gb vram) and all images are scaled properly with something like `512 x 768` resolution (they are not the same). Also I have not noticed any memory leakage with batch processing.
Can you also share some step-by-step example, cuz it’s not quite clear for me what can cause such behavior? It might be useful here as well :)

xianienie commented:
Hi
i'd like to read prompts from a txt file for image batch processing, which is very useful for video production. Are there any solutions available for this?
I have used many extensions, but still cannot generate images of different sizes.
